modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPTaxCategoryLocalServiceImpl.java

```diff
@@ -157,6 +189,93 @@ public CPTaxCategory updateCPTaxCategory(
 		return cpTaxCategoryPersistence.update(cpTaxCategory);
 	}
 
+	private SearchContext _buildSearchContext(
+		long companyId, String keywords, int start, int end, Sort sort) {
+
+		SearchContext searchContext = new SearchContext();
+
+		searchContext.setAttributes(
+			HashMapBuilder.<String, Serializable>put(
+				Field.DESCRIPTION, keywords
+			).put(
+				Field.ENTRY_CLASS_PK, keywords
+			).put(
+				Field.NAME, keywords
+			).put(
+				"params",
+				LinkedHashMapBuilder.<String, Object>put(
+					"keywords", keywords
+				).build()
+			).build());
+		searchContext.setCompanyId(companyId);
+		searchContext.setEnd(end);
+
+		if (Validator.isNotNull(keywords)) {
+			searchContext.setKeywords(keywords);
+		}
+
+		if (sort != null) {
+			searchContext.setSorts(sort);
+		}
+
+		searchContext.setStart(start);
+
+		QueryConfig queryConfig = searchContext.getQueryConfig();
+
+		queryConfig.setHighlightEnabled(false);
+		queryConfig.setScoreEnabled(false);
+
+		return searchContext;
+	}
+
+	private List<CPTaxCategory> _getCPTaxCategories(Hits hits)
+		throws PortalException {
+
+		List<Document> documents = hits.toList();
+
+		List<CPTaxCategory> cpTaxCategories = new ArrayList<>(documents.size());
+
+		for (Document document : documents) {
+			long cpTaxCategorieId = GetterUtil.getLong(
+				document.get(Field.ENTRY_CLASS_PK));
+
+			CPTaxCategory cpTaxCategory = fetchCPTaxCategory(cpTaxCategorieId);
+
+			if (cpTaxCategory == null) {
+				Indexer<CPTaxCategory> indexer = IndexerRegistryUtil.getIndexer(
+					CPTaxCategory.class);
+
+				long companyId = GetterUtil.getLong(
+					document.get(Field.COMPANY_ID));
+
+				indexer.delete(companyId, document.getUID());
+			}
+			else if (cpTaxCategory != null) {
+				cpTaxCategories.add(cpTaxCategory);
+			}
+		}
+
+		return cpTaxCategories;
+	}
+
+	private BaseModelSearchResult<CPTaxCategory> _searchCPTaxCategories(
+			SearchContext searchContext)
+		throws PortalException {
+
+		Indexer<CPTaxCategory> indexer = IndexerRegistryUtil.nullSafeGetIndexer(
```

Comment:
@MahmoudAzzam I think we should start using the searcher pattern like here:
com.liferay.account.service.impl.AccountEntryLocalServiceImpl#searchAccountEntries

It's just a smarter way of doing it

cc @andrea-ale-sbarra